src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -215,40 +216,16 @@ internal void DefineType(string script, ParserOptions parserOptions)
             foreach (var defType in definedTypes)
             {
                 var name = defType.Ident.Name.Value;
-                var res = DTypeFromTexlNode(defType.Type.TypeRoot) ?? throw new Exception("Failed defining type");
-                _definedTypeSymbolTable.RegisterType(name, new KnownRecordType(res));
-            }
-        }
-
-        internal DType DTypeFromTexlNode(TexlNode node)
-        {
-            if (node is RecordNode recordNode)
-            {
-                return DTypeFromRecordNode(recordNode);
-            }
-            else if (node is FirstNameNode nameNode)
-            {
-                return GetTypeFromName(nameNode.Ident.Name.Value);
-            }
-
-            return null;
-        }
-
-        internal DType DTypeFromRecordNode(RecordNode recordNode)
-        {
-            var list = new List<TypedName>();
-            foreach (var (node, ident) in recordNode.ChildNodes.Zip(recordNode.Ids, (a, b) => (a, b)))
-            {
-                var ty = DTypeFromTexlNode(node);
-                if (ty == null)
+                var res = defType.Type.TypeRoot.Accept(new DTypeVisitor(), _definedTypeSymbolTable) ?? throw new Exception("Failed defining type");
+                if (res.IsRecord)

```

Comment:
I think you might be able to just call `FormulaType.Build` on this result, rather than checking table/record/...